### PR TITLE
add implementation for stepwise trend in Python

### PR DESCRIPTION
**Pull request by Thuener**
Trying to fix issue #1789

I still need to do: 

- [x] test unix prophet.stan file 
- [x] find a volunteer to do the R code part
- [ ] unit tests

### CODE REVIEW

**Comment by arid-mangoss** [comment]
Hi @Thuener, 
I would be glad to take a stab at the R version of this. Before I start, I will wait to see what @bletham thinks about this feature. If he gives his blessing then I will jump on it :)

**Comment by bletham** [comment]
Thanks for the PR, and sorry for the slow follow-up on this.

The commit is really clean and looks great. Like I said on the issue I'm really hesitant to expand the built-in trends into too large of a collection because it starts to drift the package away from the type of problems it is built around and can be confusing to the end user to understand what particular trend they might want. And there is just a really large set of trends that could be of interest. For this stepwise trend, what I'd like to do is not merge it right now but do put this PR in the documentation as the prototypical example for what needs to be done to customize the trend. If we start to see a lot of people in the future posting time series that look like they would benefit from a stepwise trend then we can revisit whether or not this should be built-in, but my sense now is that it isn't an especially common trend shape and that this PR will be much more valuable as an example for others who want to implement something from the long tail of functions.

So thanks again for the contribution, I'm going to close this and will change the documentation to point to this PR as the example for adding a new trend.

**Comment by blunghino** [comment]
found my way here from the docs. stepwise trend would have been a useful feature for me!

**Comment by rbagd** [comment]
I also just came around this PR as we were thinking of submitting a similar PR with our own implementation of a step-wise trend.

@bletham, would you reconsider adding (piecewise constant | flat discontinuous | step-wise) trend into the main package? I think it's a type which is common enough to reappear across datasets and industries and now I find that we wasted hours into reinventing the wheel. 😉

As counter-arguments to your reasoning, I would say

- Current implementation of `growth="flat"` is the only non-piecewise type of trend there is and I'd even argue the piecewise version should be the default choice for `growth="flat"` for more coherence across all trend types and their link with changepoints. You can always simulate a global flat trend by disabling changepoints, so it's a little bit of a weird situation.
- Maintaining custom Prophet forks is really challenging over time.

**Comment by salva** [comment]
Hi, I am using Prophet for predicting sales of garments in a retail company and the trend for that is mostly stepwise as the mayor factor is the popularity of the collection and that factor usually remains constant along every season.

So, add this feature, please!!!

File: python/fbprophet/forecaster.py [hunk 1]
```diff
@@ -1295,6 +1344,44 @@ def flat_trend(t, m):
         """
         m_t = m * np.ones_like(t)
         return m_t
+    
+    @staticmethod
+    def stepwise(t, deltas, k, m, changepoint_ts):
```
Comment (mycaule) [inline python/fbprophet/forecaster.py:1349]: `deltas` and `k` actually unused here

File: python/fbprophet/forecaster.py [hunk 2]
```diff
@@ -1142,9 +1179,13 @@ def fit(self, df, **kwargs):
         elif self.growth == 'flat':
             dat['cap'] = np.zeros(self.history.shape[0])
             kinit = self.flat_growth_init(history)
-        else:
+        elif self.growth == 'logistic':
```
Comment (mycaule) [inline python/fbprophet/forecaster.py:1182]: should be `'logistic'` by default.

**Comment by andreastn87** [comment]
I am currently working with prophet and a stepwise built-in trend would be really helpfull for me too!

**Pull request closed without merging**